### PR TITLE
Fix the history behavior (#94)

### LINKED ISSUE #94: [Bug] History command disappear after few seconds + cursor position not updated with history command len
Step to reproduce:
- Checkout main
- Cargo run 
- Type a command
- Use low arrow to navigate the history
- Wait few seconds

Registrazione schermo 2021-07-09 alle 22 04 12

### PR DESCRIPTION
**Pull request by sholderbach**
- [x] Repaint mode switching after the 1sec case
- [x] Back to history cursor with len for not used -> fixes the skipping the
most recent
- [x] Traversal and prefix search now work as expected
- [x] Ctrl-R working (not super elegant but browsable)

Closes #94

### CODE REVIEW

**Comment by mzanrosso** [comment]
@sholderbach - I'm glad someone is fixing it, tell me if I can help you in any way

**Comment by mzanrosso** [comment]
@sholderbach - maybe it's useful to integrate with this attempt to fix by sherubthakur https://github.com/sherubthakur/reedline/commits/history-fix-cursor-update

**Comment by sholderbach** [comment]
I rebased onto that. But stepped onto new weird interactions

File: src/history/file_backed.rs [hunk 1]
```diff
@@ -218,50 +221,34 @@ impl FileBackedHistory {
...
+            if let Some((next_cursor, _)) = self
+                .entries
+                .iter()
+                .take(self.cursor)
+                .enumerate()
+                .rev()
+                .find(|x| criteria(x.1))
+            {
+                // set to entry
+                self.cursor = next_cursor
```
Comment (sherubthakur) [inline src/history/file_backed.rs:234]: This is so much better than what I had. Awesome job!

File: src/history/file_backed.rs [hunk 2]
```diff
@@ -218,50 +221,34 @@ impl FileBackedHistory {
...
-                    break;
-                }
+        if !self.entries.is_empty() {
+            if let Some((next_cursor, _)) = self
+                .entries
+                .iter()
+                .take(self.cursor)
+                .enumerate()
+                .rev()
+                .find(|x| criteria(x.1))
```
Comment (sherubthakur) [inline src/history/file_backed.rs:231]: nit-pick: I find `.find(|(_idx, string)| criteria(string)` more readable than this. Thoughts?

File: src/history/base.rs [hunk 3]
```diff
@@ -1,8 +1,10 @@
 use std::collections::vec_deque::Iter;
 
+use crate::line_buffer::LineBuffer;
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum HistoryNavigationQuery {
-    Normal,
+    Normal(LineBuffer),
```
Comment (sherubthakur) [inline src/history/base.rs:7]: Do we really want the HistoryNavigation to be aware of `LineBuffer`? in my opinion `HistoryNavigation` stuff should be able to exist on it's own without any dependency on anything related to Reedline. In a sense that it is an independent module that any line editor can use.

File: src/history/file_backed.rs [hunk 4]
```diff
@@ -89,8 +91,8 @@ impl HistoryView for FileBackedHistory {
 
     fn forward(&mut self) {
         match self.query.clone() {
-            HistoryNavigationQuery::Normal => {
-                if (self.cursor as isize) < self.entries.len() as isize - 1 {
+            HistoryNavigationQuery::Normal(_) => {
+                if self.cursor < self.entries.len() {
```
Comment (sherubthakur) [inline src/history/file_backed.rs:95]: let's say we have a vector of length 3, then the index value `3` is invalid. Right? We only have 0, 1, 2.

So shouldn't this be `self.entries.len() - 1` and not `self.entries.len()`?

Comment (sholderbach) [reply src/history/file_backed.rs:95]: We use the state `self.cursor == self.entries.len()` to denote we are not currently browsing the history. Thus this allows to leave the history and also fixes the issues of skipping the most recent entry on the first go. This is safe as `VecDeque::get()` returns `None` for the resulting out of bound access, which is desired to denote leaving the list of history entries.

File: src/history/base.rs (hunk 3 above)
Comment (sholderbach) [reply src/history/base.rs:7]: Sub-crate level modularity is currently not high on the list of design requirements. JT and I talked about this being probably the most elegant way to get the `bash` style recovery of the previously typed line buffer (if your cursor is not at the end for prefix search). This way we avoid additional state in the engine that might get out of sync. Also like this, we have a single authoritative line buffer in the `Reedline` engine all the other plugins can rely on. We probably don't have unlimited mental "state budget" for the engine, so we cheated that into this enum.

File: src/history/file_backed.rs (hunk 2 above)
Comment (sholderbach) [reply src/history/file_backed.rs:231]: Will be necessary anyways to fix the consecutive match behavior back in ;)

**Comment by sholderbach** [comment]
> I rebased onto that. But stepped onto new weird interactions

Post-mortem on that one: 

The interaction observed that reset the history browsing all the time was the introduction of the separate `Up` and `Down` edit commands for multiline, that were missed by the guard at the beginning of `run_edit_commands`

**Comment by jntrnr** [comment]
Looks good!

File: src/history/base.rs (hunk 3 above)
Comment (sherubthakur) [reply src/history/base.rs:7]: > Sub-crate level modularity is currently not high on the list of design requirements.

Ok, if by sub-carte level modulatiry you mean modular design inside this crate then won't you say that this has the potential to hinder the development later down the road?
If you were hinting about history as a separate create then I did not mean it that way, I was just using that as an example on how I think we should envision the history module API.

Although I would say that I am not really sure what is the exact problem we are trying to solve here and if there even is a better way to do so.

**Pull request merged**